aiida/engine/daemon/execmanager.py

```diff
@@ -394,6 +394,58 @@ def retrieve_calculation(calculation, transport, retrieved_temporary_folder):
     retrieved_files.add_incoming(calculation, link_type=LinkType.CREATE, link_label=calculation.link_label_retrieved)
 
 
+def stash_calculation(calculation, transport):
+    """Stash files from the working directory of a completed calculation to a permanent remote folder.
+
+    After a calculation has been completed, optionally stash files from the work directory to a storage location on the
+    same remote machine. This is useful if one wants to keep certain files from a completed calculation to be removed
+    from the scratch directory, because they are necessary for restarts, but that are too heavy to retrieve.
+    Instructions of which files to copy where are retrieved from the `stash.source_list` option.
+
+    :param calculation: the calculation job node.
+    :param transport: an already opened transport.
+    """
+    from aiida.common.datastructures import StashMode
+    from aiida.orm import RemoteData
+
+    logger_extra = get_dblogger_extra(calculation)
+
+    stash_options = calculation.get_option('stash')
+    stash_mode = stash_options.get('mode', StashMode.COPY.value)
+    source_list = stash_options.get('source_list', [])
+
+    if not source_list:
+        return
+
+    if stash_mode != StashMode.COPY.value:
+        execlogger.warning(f'stashing mode {stash_mode} is not implemented yet.')
+        return
+
+    execlogger.debug(f'stashing files for calculation<{calculation.pk}>: {source_list}', extra=logger_extra)
+
+    uuid = calculation.uuid
+    target_basepath = os.path.join(stash_options['path'], uuid[:2], uuid[2:4], uuid[4:])
+
+    for source_filename in source_list:
+
+        source_filepath = os.path.join(calculation.get_remote_workdir(), source_filename)
+        target_filepath = os.path.join(target_basepath, source_filename)
+
+        # If the source file is in a (nested) directory, create those directories first in the target directory
+        target_dirname = os.path.dirname(target_filepath)
```

Comment:
General comment - we should be careful on using os.path.* methods on remote paths, as the FS might be different. I think anyway we're doing it anywhere, so OK to keep it as it is for now.

Reply:
Good point. Surprised this has never caused problems. Wouldn't this break when you run AiiDA on windows and submit to a linux based computer? For example the `upload_calculation` function would be formatting paths to be created on the remote using windows path separators. Guess no one has ever used a Windows client or a Windows-based cluster :sweat_smile:

Reply:
I guess we suggest to use the WSL in windows so it's still using unix-like paths (I think?) and I never saw a Windows supercomputer (where I think anyway many of the actual codes might also break :-) )